sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpMessageExtensions.cs

```diff
@@ -0,0 +1,106 @@
+        {
+            ReadOnlyMemory<byte> bodyBytes = message.AmqpData;
+            var body = new ArraySegment<byte>(bodyBytes.IsEmpty ? Array.Empty<byte>() : bodyBytes.ToArray());
+            return AmqpMessage.Create(new Data { Value = body });
+        }
+
+        private static AmqpMessage CreateAmqpSequenceMessage(ServiceBusMessage message) =>
+            AmqpMessage.Create(message.AmqpSequence.Select(s => new AmqpSequence(s)));
+
+        private static AmqpMessage CreateAmqpValueMessage(ServiceBusMessage message) =>
+            AmqpMessage.Create(new AmqpValue { Value = message.AmqpValue });
+
+        public static AmqpMessage CreateAmqpMessage(this ServiceBusMessage message) =>
+            message.AmqpBodyTypeHint switch
+            {
+                ServiceBusMessage.AmqpBodyType.Sequence => CreateAmqpSequenceMessage(message),
+                ServiceBusMessage.AmqpBodyType.Value => CreateAmqpValueMessage(message),
+                // Both Data and Unspecified
+                _ => CreateAmqpDataMessage(message)
+            };
+
+        public static ServiceBusMessage.AmqpBodyType GetBodyType(this AmqpMessage message)
+        {
+            if ((message.BodyType & SectionFlag.Data) != 0 && message.DataBody != null)
+            {
+                return ServiceBusMessage.AmqpBodyType.Data;
+            }
+
+            if ((message.BodyType & SectionFlag.AmqpSequence) != 0 && message.SequenceBody != null)
+            {
+                return ServiceBusMessage.AmqpBodyType.Sequence;
+            }
+
+            if ((message.BodyType & SectionFlag.AmqpValue) != 0 && message.ValueBody.Value != null)
+            {
+                return ServiceBusMessage.AmqpBodyType.Value;
+            }
+
+            return ServiceBusMessage.AmqpBodyType.Unspecified;
+        }
+
+        private static ServiceBusReceivedMessage CreateServiceBusDataMessage(AmqpMessage message)
+        {
+            var dataSegments = new List<byte>();
+            foreach (var data in message.DataBody)
+            {
+                if (data.Value is byte[] byteArrayValue)
+                {
+                    dataSegments.AddRange(byteArrayValue);
+                }
+                else if (data.Value is ArraySegment<byte> arraySegmentValue)
+                {
+                    byte[] byteArray;
+                    if (arraySegmentValue.Count == arraySegmentValue.Array?.Length)
+                    {
+                        byteArray = arraySegmentValue.Array;
+                    }
+                    else
+                    {
+                        byteArray = new byte[arraySegmentValue.Count];
+                        Array.ConstrainedCopy(arraySegmentValue.Array, arraySegmentValue.Offset, byteArray, 0, arraySegmentValue.Count);
+                    }
+                    dataSegments.AddRange(byteArray);
+                }
+            }
+            return new ServiceBusReceivedMessage(dataSegments.ToArray());
```

Comment:
This conversion code is unchanged from the original method.

Reply:
Fixed one issue where I was using the normal constructor for this instead of the extensions method so the body type is set properly.